### PR TITLE
add a podman-compose command

### PR DESCRIPTION
**Pull request by vrothberg**
**podman compose** is a thin wrapper around an external compose provider such as docker-compose or podman-compose.  This means that `podman compose` is executing another tool that implements the compose functionality but sets up the environment in a way to let the compose provider communicate transparently with the local Podman socket.  The specified options as well the command and argument are passed directly to the compose provider.

The default compose providers are `docker-compose` and `podman-compose`. If installed, `docker-compose` takes precedence since it is the original implementation of the Compose specification and is widely used on the supported platforms (i.e., Linux, Mac OS, Windows).

If you want to change the default behavior or have a custom installation path for your provider of choice, please change the `compose_provider` field in `containers.conf(5)`.  You may also set the `PODMAN_COMPOSE_PROVIDER` environment variable.

#### Does this PR introduce a user-facing change?

```release-note
Add a new `podman compose` as a thin wrapper around an external compose provider such as docker-compose or podman-compose
```

### CODE REVIEW

**Comment by vrothberg** [comment]
@ashley-cui @Luap99  @rhatdan @benoitf PTAL
@edsantiago can you check the tests?

File: cmd/podman/compose.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+			// If no default connection is set on Linux or FreeBSD,
+			// we just use the local socket by default - just as
+			// the remote client does.
+			case "linux", "freebsd":
+				return registry.DefaultAPIAddress(), nil
+			// If there is no default connection on Windows or Mac
+			// OS, we can safely assume that something went wrong.
+			// A `podman machine init` will set the connection.
+			default:
+				return "", fmt.Errorf("cannot connect to a socket or via SSH: no default connection found")
```
Comment (mheon) [inline cmd/podman/compose.go:109]: Since we're already switching on GOOS here, it feels like we can safely recommend `podman machine init` to get a VM running in this error

**Comment by vrothberg** [comment]
Compose tests pass locally, so I guess it's installed at another path.  Will look into it soon.  For now collecting feedback and ideas.

File: go.mod [hunk 2]
```diff
@@ -188,12 +188,14 @@ require (
...
+	google.golang.org/grpc v1.56.2 // indirect
 	gopkg.in/go-jose/go-jose.v2 v2.6.1 // indirect
 	gopkg.in/square/go-jose.v2 v2.6.0 // indirect
 	gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 )
 
 replace github.com/opencontainers/runc => github.com/opencontainers/runc v1.1.1-0.20220617142545-8b9452f75cbc
+
+replace github.com/containers/common => github.com/vrothberg/common v0.0.3-0.20230720135954-8881c20f1d4e
```
Comment (edsantiago) [inline go.mod:201]: Leftover from your dev?

Comment (vrothberg) [reply go.mod:201]: No, I am vendoring the not-yet-merged common PR https://github.com/containers/common/pull/1571.  Once that PR is in, I can properly vendor.

I marked this PR here as draft to avoid accidental merges.

**Comment by mheon** [comment]
Should we try `docker compose` as well, if we can find a valid Docker?

Minus that nit, LGTM

File: docs/source/markdown/podman-compose.1.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+% podman-compose 1
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply docs/source/markdown/podman-compose.1.md:1]: I was surprised as well as `make docs` generates the .md ones but all others are included by default, see https://github.com/containers/podman/tree/main/docs/source/markdown.

**Comment by vrothberg** [comment]
> Should we try docker compose as well, if we can find a valid Docker?

That would be the ultimate fallback.  At the present, I don't think it's needed since compose is a separate binary/plugin.  If that ever changes, we may need to adjust.

**Comment by edsantiago** [comment]
@vrothberg did you try running tests under `sudo`? Failures reproduce easily on my laptop.

**Comment by vrothberg** [comment]
> @vrothberg did you try running tests under `sudo`? Failures reproduce easily on my laptop.

I did not. Thanks for the pointer!

File: cmd/podman/compose.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		logrus.Debugf("Found compose provider %q", path)
+		return path, nil
+	}
+
+	return "", errors.New("no configured compose provider found on system, please refer to `man podman-compose` for details")
+}
+
+// composeEnv returns the compose-specific environment variables.
+func composeEnv() ([]string, error) {
+	hostValue, err := func() (string, error) {
```
Comment (Luap99) [inline cmd/podman/compose.go:79]: Why is this a anonymous function? Can you split this out to a proper function, I find it hard to follow the logic with such long anonymous functions.

File: cmd/podman/compose.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+var composeCommand = &cobra.Command{
+	Use:   "compose [options]",
+	Short: "Run compose workloads via an external provider such as docker-compose or podman-compose",
+	Long: `This command is a thin wrapper around an external compose provider such as docker-compose or podman-compose.  This means that podman compose is executing another tool that implements the compose functionality but sets up the environment in a way to let the compose provider communicate transparently with the local Podman socket.  The specified options as well the command and argument are passed directly to the compose provider.
+
+The default compose providers are docker-compose and podman-compose.  If installed, docker-compose takes precedence since it is the original implementation of the Compose specification and is widely used on the supported platforms (i.e., Linux, Mac OS, Windows).
+
+If you want to change the default behavior or have a custom installation path for your provider of choice, please change the compose_provider field in containers.conf(5).  You may also set PODMAN_COMPOSE_PROVIDER environment variable.`,
+	RunE:              composeMain,
+	ValidArgsFunction: completion.AutocompleteNone,
```
Comment (Luap99) [inline cmd/podman/compose.go:33]: This will not behave good with shell completion. I think we should come up with something more reasonable.
docker compose v2 is actually using cobra as well which means we could actually just pass through the hidden completion command logic, compare `docker-compose __complete ""`.
Of course that only works with docker-compose v2 but I assume that is what most would use here.

Alternatively we can try to parse --help of the compose commands but that seems way to fragile.

File: cmd/podman/compose.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+// composeProvider provides the name of or absolute path to the compose
+// provider (i.e., the external binary such as docker-compose).
+func composeProvider() (string, error) {
+	if value, ok := os.LookupEnv("PODMAN_COMPOSE_PROVIDER"); ok {
+		return value, nil
+	}
+
+	candidates := registry.PodmanConfig().ContainersConfDefaultsRO.Engine.ComposeProviders
+	if len(candidates) == 0 {
+		return "", errors.New("no compose provider specified, please refer to `man podman-compose` for details")
```
Comment (Luap99) [inline cmd/podman/compose.go:58]: Shouldn't we use `podman-compose(1)` as syntax? Remember this runs on windows and AFAIK there is no man command on windows, same for the other rplaces

**Review by Luap99 — COMMENTED** [review]
Not a proper review yet.

File: cmd/podman/compose.go (hunk 6 above)
Comment (vrothberg) [reply cmd/podman/compose.go:58]: I'll go with "documentation" instead. The absence of man pages on Windows is a good point. FWIW, man pages are currently not installed on the Mac either ...

File: cmd/podman/compose.go (hunk 5 above)
Comment (vrothberg) [reply cmd/podman/compose.go]: Added completions now as you suggested.  Here and there, I see some garbage but it's mostly useful and an improvement.  Maybe we can iron out some of the warts.

File: cmd/podman/compose.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+	if err := composeProviderExec(args, &stdout, io.Discard, false); err != nil {
+		// Ignore errors since some providers may not expose a __complete command.
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	output := strings.Split(stdout.String(), "\n")
+	if len(output) > 1 { // Drop the last items which are cobra-specific markers
+		output = output[:len(output)-2]
+	}
+	return output, cobra.ShellCompDirectiveNoFileComp
```
Comment (Luap99) [inline cmd/podman/compose.go:63]: thanks for this

you should only throw away the last line from the stdout output not the last two.
The last `ShellCompDirective...` line is written on stderr so it is not part of this string, 
then parse the last line (`:<NUM>`) and return is as `cobra.ShellCompDirective(NUM)`, that way we return the proper completion type to make path completion work.

**Comment by vrothberg** [comment]
OK, I think we're good to go.  Let's do a final review round.

@benoitf @Luap99 @mheon @ashley-cui @rhatdan

**Comment by edsantiago** [comment]
LGTM, and CI is green after three flake restarts.
